### ISSUE TITLE
Not receiving `OO` causes session to get stuck

### ISSUE THREAD

**Issue by sorenisanerd** [FGasper/zmodemjs#33]
It seems like a relatively common problem that the terminating `OO` from `sz` gets lost or misparsed or something. See all these issues:
https://github.com/sorenisanerd/gotty/issues/46
https://github.com/tsl0922/ttyd/issues/239
https://github.com/tsl0922/ttyd/issues/279
https://github.com/Eugeny/tabby/issues/6259
https://github.com/cferdinandi/tabby/issues/132
https://github.com/Eugeny/tabby/issues/5196
https://github.com/trzsz/tabby-trzsz/issues/2
https://github.com/Eugeny/tabby/issues/5132

This lets me call `.allow_missing_OO(true)` on the session object which SEEMS to make the problem go away:

```patch
diff --git a/src/zsession.js b/src/zsession.js
index 5f0b8f9..cafe57f 100644
--- a/src/zsession.js
+++ b/src/zsession.js
@@ -193,6 +193,10 @@ Zmodem.Session = class ZmodemSession extends _Eventer {
         //console.warn("Invalid first Zmodem header", hdr);
     }
 
+    allow_missing_OO(flag) {
+        this._allow_missing_OO = !!flag;
+    }
+
     /**
      * Sets the sender function that a Session object will use.
      *
@@ -549,7 +553,7 @@ Zmodem.Session.Receive = class ZmodemReceiveSession extends Zmodem.Session {
             if (this._input_buffer.length < 2) return;
 
             //if it’s OO, then set this._bytes_after_OO
-            if (Zmodem.ZMLIB.find_subarray(this._input_buffer, OVER_AND_OUT) === 0) {
+            if (this._allow_missing_OO || Zmodem.ZMLIB.find_subarray(this._input_buffer, OVER_AND_OUT) === 0) {
 
                 //This doubles as an indication that the session has ended.
                 //We need to set this right away so that handlers like
```

...but I can't reliably trigger the problem, so my confidence in this approach is not great.

Do you have thoughts on how to address this? The spec suggests that if the `OO` does not arrive, you should just move on with your life after a brief timeout, so maybe we just need to implement that timeout? Just getting stuck in this state doesn't benefit anyone.

**Comment by FGasper** [comment]
Yeah a timeout seems reasonable; all the same, the terminating OO _should_ come since this is all TCP.

**Comment by sorenisanerd** [comment]
It very much should! I have spent hours and hours trying to pinpoint how
they get lost, but to no avail.

**Comment by FGasper** [comment]
I wonder if it’s a bug in `sz`. Maybe it neglects to flush its output buffer or something.

Maybe `_consume_ZFIN()` could be where the timeout starts, at the end of which `this._on_session_end()` would be called and `_bytes_after_OO` gets set to an empty array … and if the usual OO arrives then we clear the timeout.

**Comment by sorenisanerd** [comment]
I suspected `sz`, too. I've stared at it a LOT today and it doesn't seem to
be doing anything out of the ordinary.
https://github.com/UweOhse/lrzsz/blob/master/src/lsz.c#L1855-L1857 is
pretty standard.

I haven't been able to trigger it if I run either `sz`, `bash`, or `gotty`
under `strace` :(

Either way, if `sz` is broken, we still need to support current versions of
it.

**Comment by destinytaoer** [comment]
![1685008392368-7a637501-cb3c-4302-b7a0-4372833f40ee](https://github.com/FGasper/zmodemjs/assets/27852774/0be21d57-a2eb-496e-9707-2f2cf968c383)

i have same problem. when i use `sz` to download file, sometimes it return ZFIN at the end as described in the picture above, but sometimes it doesn't return and stuck my terminal. it more frequent occurrences when the file is larger than 100MB

it stucks the tty, not my frontend terminal. maybe it is a problem of lrzsz? what can i do for it?

In my computer, it works fine. but when i use it in k8s exec api to access a pod, that error comes

![1684935844327-9eef2bcd-96ce-4553-9d22-b014088a91db](https://github.com/FGasper/zmodemjs/assets/27852774/cd082aec-6c85-42b1-9a02-0100cd9519d5)

![1684935745685-8b2da49c-de24-4c13-a77c-74842dc25e67](https://github.com/FGasper/zmodemjs/assets/27852774/64255ce1-12f9-4981-93b5-fc4c4e4b7a5f)

**Comment by sorenisanerd** [comment]
I was [directed](https://github.com/sorenisanerd/gotty/issues/46#issuecomment-3350675902) to https://github.com/tsl0922/ttyd/commit/eca54c728af672a80b3e87d3cc34ff8ad6647420 as a fix. 

@FGasper, thoughts?

**Comment by FGasper** [comment]
@sorenisanerd That patch converts the exception into a warning, but how would that affect a hang?

It seems like there are 2 issues:
1. “unexpectedness” after ZFIN … this will trigger the exception
2. blackholed OO, causing a hang

Re #1: going by the issues you’ve linked—thank you!—at first blush I’m wondering if there’s not faulty calling logic somewhere. (zmjs is, alas, all too easy to misuse.) Have you reproduced this?

Is #2 still an issue? If so, how would the patch from ttyd fix it?

**Comment by FGasper** [comment]
Going by [this](https://github.com/FGasper/zmodemjs/blob/747c258a4f87f5ed53f688548503ed62e6e02e7c/reference/zmodem.txt#L1274-L1275), a timeout would seem an appropriate fix for #2.
